compiler/rustc_middle/src/traits/mod.rs

```diff
@@ -122,46 +96,60 @@ pub struct ObligationCauseData<'tcx> {
     /// information.
     pub body_id: hir::HirId,
 
-    pub code: ObligationCauseCode<'tcx>,
+    /// `None` for `MISC_OBLIGATION_CAUSE_CODE` (a common case, occurs ~60% of
+    /// the time). `Some` otherwise.
+    code: Option<Lrc<ObligationCauseCode<'tcx>>>,
 }
 
-impl Hash for ObligationCauseData<'_> {
+// This custom hash function speeds up hashing for `Obligation` deduplication
+// greatly by only hashing the discriminant of the `ObligationCause` enum
+// instead of its contents. That shouldn't affect hash quality much since there
+// are several other fields in `Obligation` which should be unique enough,
+// especially the predicate itself which is hashed as an interned pointer. See
+// #90996.
+impl Hash for ObligationCause<'_> {
     fn hash<H: Hasher>(&self, state: &mut H) {
         self.body_id.hash(state);
         self.span.hash(state);
         std::mem::discriminant(&self.code).hash(state);
```

Comment:
This is now hashing an Option -- probably you actually want `self.code()` here? Alternatively, my guess is that hashing the option discriminant is approximately equivalent to just not hashing the code at all, so maybe we should just drop the hash of the discriminant entirely?

Reply:
Good catch. It turns out that 99.99%+ of the time the `self.code()` is a `ImplDerivedObligation`. Not sure why, but no point hashing either discriminant. I will remove it.